jetty-websocket/jetty-websocket-tests/src/test/java/org/eclipse/jetty/websocket/tests/client/ConnectFutureAbortTest.java

```diff
@@ -0,0 +1,284 @@
+                {
+                    throw new IllegalStateException(e);
+                }
+            }
+        });
+
+        CloseTrackingEndpoint clientSocket = new CloseTrackingEndpoint();
+        Future<Session> connect = client.connect(clientSocket, WSURI.toWebsocket(server.getURI()));
+
+        // Abort when session is created, this is before future has been added to session and before the connection upgrade.
+        assertTrue(enteredListener.await(5, TimeUnit.SECONDS));
+        assertTrue(connect.cancel(true));
+        assertThrows(CancellationException.class, () -> connect.get(5, TimeUnit.SECONDS));
+        exitListener.countDown();
+        assertFalse(clientSocket.openLatch.await(2, TimeUnit.SECONDS));
+
+        // Strange that onError() is called but onOpen() is never called. (This would not be the case in Jetty-10).
+        assertThat(clientSocket.error.get(), instanceOf(CancellationException.class));
+    }
+
+    @Test
+    public void testAbortInHandshakeResponse() throws Exception
+    {
+        start(c -> c.addMapping("/", EventSocket.EchoSocket.class));
+
+        final CountDownLatch enteredListener = new CountDownLatch(1);
+        final CountDownLatch exitListener = new CountDownLatch(1);
+        UpgradeListener upgradeListener = new AbstractUpgradeListener()
+        {
+            @Override
+            public void onHandshakeResponse(UpgradeResponse response)
+            {
+                try
+                {
+                    enteredListener.countDown();
+                    exitListener.await();
+                }
+                catch (InterruptedException e)
+                {
+                    throw new IllegalStateException(e);
+                }
+            }
+        };
+
+        CloseTrackingEndpoint clientSocket = new CloseTrackingEndpoint();
+        ClientUpgradeRequest upgradeRequest = new ClientUpgradeRequest();
+        Future<Session> connect = client.connect(clientSocket, WSURI.toWebsocket(server.getURI()), upgradeRequest, upgradeListener);
+
+        // Abort after after handshake response, which is before connection upgrade, but after future has been set on session.
+        assertTrue(enteredListener.await(5, TimeUnit.SECONDS));
+        assertTrue(connect.cancel(true));
+        assertThrows(CancellationException.class, () -> connect.get(5, TimeUnit.SECONDS));
+        exitListener.countDown();
+        assertFalse(clientSocket.openLatch.await(2, TimeUnit.SECONDS));
+
+        // Strange that onError() is called but onOpen() is never called. (This would not be the case in Jetty-10).
+        assertThat(clientSocket.error.get(), instanceOf(CancellationException.class));
+    }
+
+    @Test
+    public void testAbortOnOpened() throws Exception
+    {
+        start(c -> c.addMapping("/", EventSocket.EchoSocket.class));
+
+        final CountDownLatch exitOnOpen = new CountDownLatch(1);
+        CloseTrackingEndpoint clientSocket = new CloseTrackingEndpoint()
+        {
```

Comment:
This is probably going to fail in Jetty 10 since it's an anonymous endPoint and will fail the method handles permissions. Perhaps make it a proper `public` inner class now to avoid merge conflicts in the future?

Reply:
This shouldn't fail in Jetty 10, this `CloseTrackingEndpoint` extends the `WebSocketListener`. I think that issue with the public inner class is only for the annotated version as we need to discover the methods.

Reply:
Issue #5043 opened for this.